### PR TITLE
Add state to dwio common writer

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Add state to dwio common writer and apply to SortWriter and dwrf writer.
We can leverage this for common state transition handling and ease memory
arbitration related processing.

### CODE REVIEW

File: velox/dwio/common/Writer.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      }
+      break;
+    case State::kAborted:
+      [[fallthrough]];
+    case State::kClosed:
+      [[fallthrough]];
+    default:
+      break;
+  }
+  VELOX_FAIL("Unexpected state transition from {} to {}", oldState, newState);
```
Comment (mbasmanova) [inline velox/dwio/common/Writer.cpp:39]: Just to confirm, this will fail if from and to states are the same.

File: velox/dwio/common/Writer.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      VELOX_UNREACHABLE("BAD STATE: {}", static_cast<int>(state));
+  }
+}
+f bool Writer::isRunning() const {
+  return state_ == State::kRunning;
+}
+
+void Writer::checkRunning() const {
+  VELOX_CHECK_EQ(
+      state_, State::kRunning, "Writer is not in running: {}", state_);
```
Comment (mbasmanova) [inline velox/dwio/common/Writer.cpp:62]: typo: "in running" -> "running" or "in running state"

**Review by mbasmanova — APPROVED** [review]
Nice refactoring. Thanks.

File: velox/dwio/common/Writer.cpp (hunk 1 above)
Comment (xiaoxmeng) [reply velox/dwio/common/Writer.cpp:39]: Yes, it will

**Pull request closed without merging**